include/swift/AST/Types.h

```diff
@@ -3085,6 +3085,14 @@ class AnyFunctionType : public TypeBase {
       static_assert(
           ASTExtInfoBuilder::NumMaskBits == NumAFTExtInfoBits,
           "ExtInfo and AnyFunctionTypeBitfields must agree on bit size");
+
+      // Canonicalize all thin functions to be escaping (to keep compatibility
+      // with generic parameters).
+      Representation rep = Info.getValue().getRepresentation();
+      if (rep == FunctionTypeRepresentation::CFunctionPointer) {
```

Comment:
You could do `isThinRepresentation(rep)` here, which will include `Thin` representation, and any other no-context function representations we might add later.